base/libgit2/oid.jl

```diff
@@ -51,32 +61,40 @@ function GitHash(repo::GitRepo, ref_name::AbstractString)
     return oid_ptr[]
 end
 
-function GitHash(obj::Ptr{Void})
-    oid_ptr = ccall((:git_object_id, :libgit2), Ptr{UInt8}, (Ptr{Void},), obj)
-    oid_ptr == C_NULL && return GitHash()
-    return GitHash(oid_ptr)
-end
-
-function GitHash{T<:GitObject}(obj::T)
-    obj === nothing && return GitHash()
-    return GitHash(obj.ptr)
+function GitHash(obj::GitObject)
+    GitHash(ccall((:git_object_id, :libgit2), Ptr{UInt8}, (Ptr{Void},), obj.ptr))
 end
 
 Base.hex(id::GitHash) = join([hex(i,2) for i in id.val])
+Base.hex(id::GitShortHash) = hex(id.hash)[1:id.len]
 
 raw(id::GitHash) = collect(id.val)
 
-Base.string(id::GitHash) = hex(id)
+Base.string(id::AbstractGitHash) = hex(id)
 
-Base.show(io::IO, id::GitHash) = print(io, "GitHash($(string(id)))")
+Base.show(io::IO, id::GitHash) = print(io, "GitHash(\"$(string(id))\")")
+Base.show(io::IO, id::GitShortHash) = print(io, "GitShortHash(\"$(string(id))\")")
 
 Base.hash(id::GitHash, h::UInt) = hash(id.val, h)
 
-cmp(id1::GitHash, id2::GitHash) = Int(ccall((:git_oid_cmp, :libgit2), Cint,
-                                    (Ptr{GitHash}, Ptr{GitHash}), Ref(id1), Ref(id2)))
+function Base.cmp(id1::GitHash, id2::GitHash)
+    Int(ccall((:git_oid_cmp, :libgit2), Cint,
+              (Ptr{GitHash}, Ptr{GitHash}),
+              Ref(id1), Ref(id2)))
+end
+function Base.cmp(id1::GitShortHash, id2::GitShortHash)
+    # shortened hashes appear at the beginning of the order, i.e.
+    # 000 < 01 < 010 < 011 < 0112
+    c = Int(ccall((:git_oid_ncmp, :libgit2), Cint,
+                  (Ptr{GitHash}, Ptr{GitHash}, Csize_t),
+                  Ref(id1.hash), Ref(id2.hash), min(id1.len, id2.len)))
+    return c == 0 ? cmp(id1.len, id2.len) : c
+end
+Base.cmp(id1::GitHash, id2::GitShortHash) = cmp(GitShortHash(id1, OID_HEXSZ), id2)
+Base.cmp(id1::GitShortHash, id2::GitHash) = cmp(id1, GitShortHash(id2, OID_HEXSZ))
 
-==(id1::GitHash, id2::GitHash) = cmp(id1, id2) == 0
-Base.isless(id1::GitHash, id2::GitHash)  = cmp(id1, id2) < 0
+==(id1::AbstractGitHash, id2::AbstractGitHash) = cmp(id1, id2) == 0
```

Comment:
two references to the same object comparing as not equal because they had different partial hash lengths strikes me as unintuitive and not a useful notion of equality